### PR TITLE
Fine-tuning of the OpenGL state

### PR DESCRIPTION
**Pull request by devmessias**
This it's related with  #428

Hi @Garyfallidis and @skoudoro. I didn't finish yet, but I'm sending the pull-request to get feedbacks or suggestions.

I've created a function inside of shaders/base.py (shader_apply_effects) in order to make the control of the opengl_state more  simple.

```python
id_observer = shader_apply_effects(
    showm, actorAdd,
    effect=window.gl_set_additive_blending)

###############################################################################
# It's also possible to pass a list of effects. The final opengl state it'll  be the composition of each 
# effect that each function has in the opengl state

id_observer =  shader_apply_effects(
    showm, actorNoDepthTest2,
    effects=[
        window.gl_reset_blend, window.gl_disable_blend,
        window.gl_disable_depth])
 ```

### CODE REVIEW

**Review by filipinascimento — COMMENTED** [review]
Please, can you add a few more details about the types of effects you can attain using this approach? In particular, can you elaborate (maybe give some screenshots) of the effects of disabling depth buffer and playing with different blending techniques?

File: fury/shaders/base.py [hunk 1]
```diff
@@ -200,6 +201,50 @@ def cbk(caller, event, calldata=None):
...
+            can be override by the others.
+
+    Returns:
+    --------
+        id_observer : int
+            An unsigned Int tag which can be used later to remove the event
+            or retrieve the vtkCommand used in the observer.
+            See more at: https://vtk.org/doc/nightly/html/classvtkObject.html
+
+    """
```
[the remark this answers is not in the document]
Comment (devmessias) [reply fury/shaders/base.py:228]: I agree. Maybe apply_gl_context

File: docs/examples/viz_fine_tuning_gl_context.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+centers_no_depth_test2 = centers_additive - np.array([[0, -1, 0]])
+colors = np.array([
+    [1, 0, 0],
+    [0, 1, 0],
+    [0, 0, 1]
+])
+
+
+useSDFActors = True
+if useSDFActors:
```
[the remark this answers is not in the document]
Comment (devmessias) [reply docs/examples/viz_fine_tuning_gl_context.py:46]: Spheres are "pure" vtk objects. There are  other effects and specific opengl states being applied in the vtk spheres. Shader actors are more suitable to test the true effects of a specific opengl context.

File: docs/examples/viz_fine_tuning_gl_context.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+# inside of a shader callback, even if you have just one actor.
+
+glState = showm.window.GetState()
+
+
+def gl_disable_depth(glState):
+    '''this functions it's allways accessible through
+    fury.window.gl_disable_depth
+    '''
+    GL_DEPTH_TEST = 2929
```
[the remark this answers is not in the document]
Comment (devmessias) [reply docs/examples/viz_fine_tuning_gl_context.py:96]: http://oss.infoscience.co.jp/factor/docs.factorcode.org/content/word-GL_BLEND,opengl.gl.html

File: docs/examples/viz_fine_tuning_gl_context.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+###############################################################################
+# It's also possible to pass a list of effects. The final opengl state it'll 
+# be the composition of each effect that each function has in the opengl state
+
+shader_apply_effects(
+    showm, actorNoDepthTest2,
+    effects=[
+        window.gl_reset_blend, window.gl_disable_blend,
+        window.gl_disable_depth])
+
```
[the remark this answers is not in the document]
Comment (devmessias) [reply docs/examples/viz_fine_tuning_gl_context.py:143]: No, you don't need. But this is just an example about how to create custom behavior from scratch. Also, it gives you an idea of what's going on behind the doors.

File: docs/examples/viz_fine_tuning_gl_context.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+def gl_disable_depth(glState):
+    '''this functions it's allways accessible through
+    fury.window.gl_disable_depth
+    '''
+    GL_DEPTH_TEST = 2929
+    GL_BLEND = 3042
+    glState.vtkglDisable(GL_DEPTH_TEST)
+    glState.vtkglDisable(GL_BLEND)
```
[the remark this answers is not in the document]
Comment (devmessias) [reply docs/examples/viz_fine_tuning_gl_context.py:99]: It's just an example with more options for the reader. See the gl_disable_depth  implementation https://github.com/devmessias/fury/blob/aeb2596c75c9490de6afc660d1f7e1ceeb505987/fury/window.py#L1228

**Comment by Garyfallidis** [comment]
Hi @devmessias this an interesting approach. This PR needs some cleaning up. Effects is not so useful name. Because effects could be anything. I suggest stick to state or `gl_state` or `graphical_state`.

**Comment by Garyfallidis** [comment]
Okay I understand what you want to do. Will provide step by step comments tomorrow (or asap). You can start working on the tests.

File: docs/examples/viz_fine_tuning_gl_context.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+from fury import window, actor
+import itertools
+
+
+###############################################################################
+# We just proceeds as usual: creating the actors and initializing a scene in
+# FURY
+
+centers = 1*np.array([
```
Comment (skoudoro) [inline docs/examples/viz_fine_tuning_gl_context.py:28]: You can remove this multiplication by 1

File: docs/examples/viz_fine_tuning_gl_context.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+                           size=(900, 768), reset_camera=False,
+                           order_transparent=True)
+
+###############################################################################
+# All actors must be added  in the scene
+
+scene.add(actors)
+scene.add(actor_add)
+scene.add(actors_no_depth_test)
+###############################################################################
```
Comment (skoudoro) [inline docs/examples/viz_fine_tuning_gl_context.py:61]: missing the empty line above. Otherwise, the documentation will not render correctly

File: docs/examples/viz_fine_tuning_gl_context.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    cnt = next(counter)
+    showm.render()
+    if cnt == 100:
+        actor_add.GetMapper().RemoveObserver(id_add)
+    if cnt == 1000:
+        showm.exit()
+
+
+showm.add_timer_callback(True, 5, timer_callback)
+showm.start()
```
Comment (skoudoro) [inline docs/examples/viz_fine_tuning_gl_context.py:108]: missing  `window.render` for the illustration

File: fury/shaders/base.py [hunk 9]
```diff
@@ -200,6 +221,50 @@ def cbk(caller, event, calldata=None):
...
+def shader_apply_effects(
+        showm, actor, effect=None, effects=None, priority=0):
+    """This applies a specific opengl state (effect) or a list of effects just
+    before the actor's shader be executed.
+
+    Parameters
+    ----------
+    showm : fury.window.ShowManager
+    actor : vtk actor
+    effect :  function
```
Comment (skoudoro) [inline fury/shaders/base.py:233]: `, optional`

File: fury/shaders/base.py [hunk 10]
```diff
@@ -200,6 +221,50 @@ def cbk(caller, event, calldata=None):
...
+        a function with a glState as argument
+    effects : a list of functions, optional
```
Comment (skoudoro) [inline fury/shaders/base.py:235]: confusing to have `effects` and `effect`. I believe it can be only one argument. You can use `isinstance(object, Iterable) or isinstance(object, Callable)` to difference them

File: fury/shaders/base.py [hunk 11]
```diff
@@ -200,6 +221,50 @@ def cbk(caller, event, calldata=None):
...
+    -------
+    id_observer : int
+        An unsigned Int tag which can be used later to remove the event
+        or retrieve the vtkCommand used in the observer.
+        See more at: https://vtk.org/doc/nightly/html/classvtkObject.html
+
+    """
+    if effects is None:
+        effects = [effect]
+
```
Comment (skoudoro) [inline fury/shaders/base.py:251]: What happens when both (`effect` and `effects`) are defined?

File: fury/shaders/base.py [hunk 12]
```diff
@@ -200,6 +221,50 @@ def cbk(caller, event, calldata=None):
...
+    def callback(
+            _caller, _event, calldata=None,
+            effects=[], showm=None):
```
Comment (skoudoro) [inline fury/shaders/base.py:254]: you should avoid `[]` as the default parameter.

For your information: https://docs.python-guide.org/writing/gotchas/

File: fury/window.py [hunk 13]
```diff
@@ -1146,3 +1146,107 @@ def enable_stereo(renwin, stereo_type):
         stereo_type = 'horizontal'
 
     renwin.SetStereoType(stereo_type_dictionary[stereo_type])
+
+
+def test_and_extract_gl_state(func):
```
Comment (skoudoro) [inline fury/window.py:1151]: Can you just rename it `extract_gl_state`

File: fury/window.py [hunk 14]
```diff
@@ -1146,3 +1146,107 @@ def enable_stereo(renwin, stereo_type):
...
+        """
+        if isinstance(obj, ShowManager):
+            glState = obj.window.GetState()
+
+        elif not isinstance(obj, vtk.vtkOpenGLState):
+            glState = obj
+
+        else:
+            raise TypeError("""valid types are vtkOpenGLState
+             or fury.window.ShowManager """)
```
Comment (skoudoro) [inline fury/window.py:1168]: indentation strange

**Review by skoudoro — COMMENTED** [review]
Hi @devmessias,

Thank you for this. Can you add some tests? 

Overall it looks good. 

I still wonder if the function in the `window` module should be somewhere else. I need to think about it.

**Comment by skoudoro** [comment]
Also, you might need to rebase this PR.

File: fury/shaders/base.py (hunk 10 above)
Comment (devmessias) [reply fury/shaders/base.py:235]: Ok. I think that it's better.

Comment (devmessias) [reply fury/shaders/base.py]: Done

File: fury/window.py (hunk 13 above)
Comment (devmessias) [reply fury/window.py]: @Garyfallidis  and @filipinascimento  asked to me to remove this decorator behavior.

Comment (skoudoro) [reply fury/window.py]: ok

File: docs/examples/viz_fine_tuning_gl_context.py [hunk 15]
```diff
@@ -0,0 +1,160 @@
+"""
+=======================================================
+Fine-tuning the OpenGL state using shader callbacks
+=======================================================
+
+VTK it’s powerfully, but sometimes we need to get more control about how
```
Comment (Garyfallidis) [inline docs/examples/viz_fine_tuning_gl_context.py:6]: Remove from "VTK to specific actor".

Switch to: 
This example shows how to change the OpenGL state of one or more actors.

File: docs/examples/viz_fine_tuning_gl_context.py [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+OpenGL will render the actors.  For example, enforcing that deep-test keep
+disabled during the draw call of a specific actor.  This can be useful when
+we need to fine-tuning the performance or create specific visualization
```
Comment (Garyfallidis) [inline docs/examples/viz_fine_tuning_gl_context.py:9]: Remove we need to fine-tuning the performance. Say...

This can be useful when we need to create specialized visualization effects.

File: docs/examples/viz_fine_tuning_gl_context.py [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+import numpy as np
+
+from fury.shaders import shader_apply_effects
+
+from fury import window, actor
+import itertools
+
+
+###############################################################################
+# We just proceeds as usual: creating the actors and initializing a scene in
```
Comment (Garyfallidis) [inline docs/examples/viz_fine_tuning_gl_context.py:26]: proceed

File: docs/examples/viz_fine_tuning_gl_context.py [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+
+# After some steps we will remove the no_depth_test effect
+
+
+def timer_callback(obj, event):
+    cnt = next(counter)
+    showm.render()
+    showm.scene.GetActiveCamera().Azimuth(1)
+    if cnt == 400:
+        actor_no_depth_test.GetMapper().RemoveObserver(id_observer)
```
Comment (Garyfallidis) [inline docs/examples/viz_fine_tuning_gl_context.py:144]: Why is this necessary to call here inside the time_callback?

File: fury/shaders/base.py [hunk 19]
```diff
@@ -200,6 +221,49 @@ def cbk(caller, event, calldata=None):
...
 
 
+def shader_apply_effects(
+        window, actor, effects, priority=0):
+    """This applies a specific opengl state (effect) or a list of effects just
+    before the actor's shader be executed.
+
+    Parameters
+    ----------
+    window : vtk.vtkRenderWindow
```
Comment (Garyfallidis) [inline fury/shaders/base.py:231]: window : ShowManager.win

File: fury/shaders/base.py [hunk 20]
```diff
@@ -200,6 +221,49 @@ def cbk(caller, event, calldata=None):
...
+    actor : vtk actor
```
Comment (Garyfallidis) [inline fury/shaders/base.py:232]: actor : actor

**Review by Garyfallidis — COMMENTED** [review]
A lot of code not covered by tests. Is there a reason for that?

**Review by Garyfallidis — CHANGES_REQUESTED** [review]
More comments. See above.

**Comment by skoudoro** [comment]
Hi @devmessias,

Can you address all the comments above before the next meeting? 

It would be good to have this PR merged

**Comment by devmessias** [comment]
> Hi @devmessias,
> 
> Can you address all the comments above before the next meeting?
> 
> It would be good to have this PR merged

Ok, I'll do it tomorrow.

**Review by devmessias — COMMENTED** [review]
> A lot of code not covered by tests. Is there a reason for that?

I forgot to call the Render() method after changing the opengl-state.

File: docs/examples/viz_fine_tuning_gl_context.py [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+        shader_apply_effects(
+             showm.window, actor_no_depth_test,
+             effects=window.gl_set_additive_blending)
+    if cnt == 1000:
+        showm.exit()
+
+
+showm.add_timer_callback(True, 5, timer_callback)
+
+interactive = True
```
Comment (skoudoro) [inline docs/examples/viz_fine_tuning_gl_context.py:155]: Should be `False`

File: docs/examples/viz_fine_tuning_gl_context.py [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+    [-.1, 0, 0],
+    [.1, 0, 0]
+])
+colors = np.array([
+    [1, 0, 0],
+    [0, 1, 0],
+    [0, 0, 1]
+])
+
+actor_no_depth_test = actor.markers(
```
[the remark this answers is not in the document]
Comment (Garyfallidis) [reply docs/examples/viz_fine_tuning_gl_context.py:38]: For example, if we would use ``actor.axes()`` which does not use FURY shaders would that work?

File: fury/tests/test_window.py [hunk 23]
```diff
@@ -332,3 +333,95 @@ def test_content(filename='fury.png', colors_found=(True, True)):
...
+        scene,
+        size=(900, 768), reset_camera=False,
+        order_transparent=False)
+
+    scene.add(actor_no_depth_test)
+
+    showm.render()
+    state = window.gl_get_current_state(showm.window.GetState())
+    before_depth_test = state['GL_DEPTH_TEST']
+    assert before_depth_test
```
Comment (Garyfallidis) [inline fury/tests/test_window.py:411]: We do not use ``assert`` statements like that in FURY tests. Please use numpy testing as we use in the other tests. Numpy testing is more explicit and therefore preferable. Also makes it easy to test arrays etc.

**Comment by Garyfallidis** [comment]
@devmessias I added a new series of comments. Please address the different issues.

File: docs/examples/viz_fine_tuning_gl_context.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (devmessias) [reply docs/examples/viz_fine_tuning_gl_context.py:38]: Some specific blending states may not work.  I talked with @filipinascimento  about that. There is a issue with the VTK's SetOpacity method  that brings  imcopatibility. I'll investigate this in the future

File: fury/shaders/base.py [hunk 24]
```diff
@@ -200,6 +221,50 @@ def cbk(caller, event, calldata=None):
...
+    def callback(
+            _caller, _event, calldata=None,
+            effects=None, window=None):
+        program = calldata
+        glState = window.GetState()
+        if program is not None:
+            for func in effects:
+                func(glState)
+
+    id_observer = add_shader_callback(
```
[the remark this answers is not in the document]
Comment (devmessias) [reply fury/shaders/base.py:260]: Yes, I've tested that using the markers actors. It seems ok. @filipinascimento  do you have any considerations about that?

File: fury/window.py [hunk 25]
```diff
@@ -1008,3 +1008,90 @@ def enable_stereo(renwin, stereo_type):
...
+def gl_set_normal_blending(glState):
+    glState.vtkglEnable(_GL['GL_BLEND'])
+    glState.vtkglEnable(_GL['GL_DEPTH_TEST'])
+    glState.vtkglBlendFunc(_GL['GL_ONE'], _GL['GL_ONE'])
+    glState.vtkglBlendFuncSeparate(
+                _GL['GL_SRC_ALPHA'], _GL['GL_ONE_MINUS_SRC_ALPHA'],
+                _GL['GL_ONE'], _GL['GL_ONE_MINUS_SRC_ALPHA'])
+
+
+def gl_set_multiplicative_blending(glState):
```
[the remark this answers is not in the document]
Comment (devmessias) [reply fury/window.py:1090]: There is a method called reset_blend_state.

File: docs/examples/viz_fine_tuning_gl_context.py [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+    cnt = next(counter)
+    showm.render()
+    # we will rotate the visualization just to help you to see
+    # the results of each specifc opengl-state
+    showm.scene.GetActiveCamera().Azimuth(1)
+    if cnt == 400:
+        actor_no_depth_test.GetMapper().RemoveObserver(id_observer)
+        shader_apply_effects(
+             showm.window, actor_no_depth_test,
+             effects=window.gl_set_additive_blending)
```
[the remark this answers is not in the document]
Comment (devmessias) [reply docs/examples/viz_fine_tuning_gl_context.py:147]: For blending we have reset_blend_state

**Comment by Garyfallidis** [comment]
@devmessias can you please address all last comments? Then we should be able to merge this.

File: docs/examples/viz_fine_tuning_gl_context.py (hunk 18 above)
Comment (devmessias) [reply docs/examples/viz_fine_tuning_gl_context.py]: Done

File: docs/examples/viz_fine_tuning_gl_context.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (Garyfallidis) [reply docs/examples/viz_fine_tuning_gl_context.py:147]: What about the others?

File: docs/examples/viz_fine_tuning_gl_context.py (hunk 15 above)
Comment (Garyfallidis) [reply docs/examples/viz_fine_tuning_gl_context.py:6]: Remove VTK it's powerful. Just say:
Sometimes we need to get more control on how OpenGL with render the actors.

Comment (devmessias) [reply docs/examples/viz_fine_tuning_gl_context.py]: Done

**Review by skoudoro — COMMENTED** [review]
LGTM!

@Garyfallidis, @filipinascimento, Do you have any other comments? Otherwise, I will go ahead and merge this PR.

**Comment by Garyfallidis** [comment]
Great job! Thank you @devmessias !!!

**Comment by Garyfallidis** [comment]
A note here that the failed CIs were irrelevant to this PR.
